s3datarepository/users.go

```diff
@@ -0,0 +1,447 @@
+
+	groupName := fmt.Sprintf("%s-DsTmpGrp", name)
+	usersOutput, err := s.listGroupsUsers(ctx, groupName)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("got iam users response %+v", usersOutput)
+
+	output := make(map[string]interface{}, len(usersOutput))
+	for _, u := range usersOutput {
+		keyOut, err := s.IAM.ListAccessKeysWithContext(ctx, &iam.ListAccessKeysInput{UserName: u.UserName})
+		if err != nil {
+			return nil, ErrCode("failed to getting access keys for user "+aws.StringValue(u.UserName), err)
+		}
+
+		keys := make(map[string]string, len(keyOut.AccessKeyMetadata))
+		for _, k := range keyOut.AccessKeyMetadata {
+			keys[aws.StringValue(k.AccessKeyId)] = aws.StringValue(k.Status)
+		}
+
+		output[*u.UserName] = struct {
+			Keys map[string]string `json:"keys"`
+		}{keys}
+	}
+
+	log.Debugf("returning map of users for dataset %s: %+v", id, output)
+
+	return output, nil
+}
+
+// listGroupsUsers lists the users that belong to a group
+func (s *S3Repository) listGroupsUsers(ctx context.Context, groupName string) ([]*iam.User, error) {
+	users := []*iam.User{}
+	if groupName == "" {
+		return users, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty group name"))
+	}
+
+	log.Infof("listing iam users for group %s", groupName)
+
+	input := &iam.GetGroupInput{GroupName: aws.String(groupName)}
+
+	truncated := true
+	for truncated {
+		output, err := s.IAM.GetGroupWithContext(ctx, input)
+		if err != nil {
+			return users, ErrCode("failed to getting users for group "+groupName, err)
+		}
+
+		truncated = aws.BoolValue(output.IsTruncated)
+		users = append(users, output.Users...)
+		input.Marker = output.Marker
+	}
+
+	return users, nil
+}
+
+func (s *S3Repository) CreateUser(ctx context.Context, id string) (interface{}, error) {
+	if id == "" {
+		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty id"))
+	}
+
+	name := id
+	if s.NamePrefix != "" {
+		name = s.NamePrefix + "-" + name
+	}
+
+	log.Infof("creating user of the s3datarepository %s", name)
+
+	policyDoc, err := s.temporaryAccessPolicy(id)
+	if err != nil {
+		return nil, ErrCode("generate temporary access policy for dataset "+id, err)
+	}
+
+	log.Debugf("generated temporary access policy document %s", string(policyDoc))
+
+	// setup rollback function list and defer execution
+	var rollBackTasks []func() error
+	defer func() {
+		if err != nil {
+			log.Errorf("recovering from error creating user in s3datarepository: %s, executing %d rollback tasks", err, len(rollBackTasks))
+			rollBack(&rollBackTasks)
+		}
+	}()
+
+	policyOutput, err := s.IAM.CreatePolicyWithContext(ctx, &iam.CreatePolicyInput{
+		Description:    aws.String("Temporary access policy for dataset " + id),
+		Path:           aws.String(s.IAMPathPrefix),
+		PolicyDocument: aws.String(string(policyDoc)),
+		PolicyName:     aws.String(name + "-DsTmpPlc"),
+	})
+
+	if err != nil {
+		return nil, ErrCode("create temporary access policy for dataset "+id, err)
```

Comment:
I noticed you don't used `failed to ...` in these error messages, which was one of the recommendations from that style guide, since the get wrapped. I should probably fix the ones in my code 😸

Reply:
lol yeah im trying it out 😆